File: modules/paralleization/tutorial.langchain.ts
/**
 * Run:
 * cd modules && node --experimental-strip-types ./paralleization/tutorial.langchain.ts
 *
 * 학습 목표:
 * 1) LangChain LCEL에서 프롬프트/모델/파서를 조합하는 방법
 * 2) RunnableMap으로 여러 체인을 병렬로 실행하는 방법
 * 3) 병렬 결과를 한 번에 수집해 구조화된 출력으로 다루는 방법
 */
import { StringOutputParser } from "@langchain/core/output_parsers";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { RunnableMap } from "@langchain/core/runnables";
import { ChatOpenAI } from "@langchain/openai";

import { isDirectExecution } from "./tutorial.execution.ts";
import { type Logger, type TutorialStepResult } from "./tutorial.types.ts";

const STEP_ID: string = "step-2-langchain";
const LIBRARY: "langchain" = "langchain";
const OPENAI_MODEL: string = "gpt-4o-mini";
const SYNTHETIC_MODEL: string = "hf:moonshotai/Kimi-K2.5";
const SYNTHETIC_BASE_URL: string = "https://api.synthetic.new/openai/v1";
const TUTORIAL_TOPIC: string = "Artificial Intelligence in Healthcare";

function toStepResult(
  status: TutorialStepResult["status"],
  message: string,
  output?: string,
): TutorialStepResult {
  const result: TutorialStepResult = {
    stepId: STEP_ID,
    library: LIBRARY,
    status,
    message,
    output,
  };
  return result;
}

/**
 * STEP 로그를 고정 형식으로 통일하면
 * ADK/OpenAI 튜토리얼과 같은 시선으로 비교 학습할 수 있습니다.
 */
function logStep(log: Logger, stepNo: number, title: string): void {
  log(`[langchain] STEP ${stepNo}/3 ${title}`);
}

type ModelConfig = {
  apiKey: string;
  modelName: string;
  baseUrl?: string;
};

function resolveModelConfig(): ModelConfig | undefined {
  // 우선순위 1: OPENAI_API_KEY (OpenAI 기본 경로)
  const openAiApiKey: string | undefined = process.env.OPENAI_API_KEY?.trim();
  if (openAiApiKey) {
    return {
      apiKey: openAiApiKey,
      modelName: OPENAI_MODEL,
    };
  }

  // 우선순위 2: SYNTHETIC_API_KEY (OpenAI 호환 엔드포인트 경로)
  const syntheticApiKey: string | undefined = process.env.SYNTHETIC_API_KEY?.trim();
  if (syntheticApiKey) {
    return {
      apiKey: syntheticApiKey,
      modelName: SYNTHETIC_MODEL,
      baseUrl: SYNTHETIC_BASE_URL,
    };
  }

  return undefined;
}

export async function runLangChainStep(
  log: Logger = console.log,
): Promise<TutorialStepResult> {
  // Step 1) 실행 준비: 모델/키 구성
  logStep(log, 1, "setup");
  const modelConfig: ModelConfig | undefined = resolveModelConfig();
  if (!modelConfig) {
    return toStepResult(
      "skipped",
      "OPENAI_API_KEY or SYNTHETIC_API_KEY is missing.",
    );
  }

  const model: ChatOpenAI = new ChatOpenAI({
    apiKey: modelConfig.apiKey,
    model: modelConfig.modelName,
    temperature: 0.2,
    configuration: modelConfig.baseUrl
      ? {
          baseURL: modelConfig.baseUrl,
        }
      : undefined,
  });

  // Step 2) 병렬 체인 구성
  // LCEL의 핵심: "프롬프트 -> 모델 -> 파서"를 파이프로 연결해 단일 runnable로 만든다.
  logStep(log, 2, "build parallel chains");
  const parser: StringOutputParser = new StringOutputParser();
  const summaryPrompt: ReturnType<typeof ChatPromptTemplate.fromMessages> =
    ChatPromptTemplate.fromMessages([
      ["system", "Summarize the following topic in one sentence."],
      ["user", "{topic}"],
    ]);
  const questionsPrompt: ReturnType<typeof ChatPromptTemplate.fromMessages> =
    ChatPromptTemplate.fromMessages([
      ["system", "Generate two interesting questions for this topic."],
      ["user", "{topic}"],
    ]);
  const keywordsPrompt: ReturnType<typeof ChatPromptTemplate.fromMessages> =
    ChatPromptTemplate.fromMessages([
      ["system", "Extract three key terms as comma-separated text."],
      ["user", "{topic}"],
    ]);

  const parallelMap: RunnableMap<
    {
      topic: string;
    },
    {
      summary: string;
      questions: string;
      keywords: string;
    }
  > = RunnableMap.from({
    // 세 체인이 같은 입력(topic)을 받아 동시에 실행된다.
    summary: summaryPrompt.pipe(model).pipe(parser),
    questions: questionsPrompt.pipe(model).pipe(parser),
    keywords: keywordsPrompt.pipe(model).pipe(parser),
  });

  // Step 3) 병렬 실행 결과 수집
  logStep(log, 3, "execute parallel map");
  const result: {
    summary: string;
    questions: string;
    keywords: string;
  } = await parallelMap.invoke({
    topic: TUTORIAL_TOPIC,
  });

  const output: string = JSON.stringify(result, null, 2);
  return toStepResult("completed", "LangChain tutorial completed.", output);
}

const isNodeTestContext: boolean = process.env.NODE_TEST_CONTEXT !== undefined;
if (!isNodeTestContext && isDirectExecution(import.meta.url, process.argv)) {
  void runLangChainStep().then((result: TutorialStepResult): void => {
    console.log(JSON.stringify(result, null, 2));
  }).catch((error: unknown): void => {
    const errorMessage: string =
      error instanceof Error ? error.message : String(error);
    console.error(errorMessage);
    process.exitCode = 1;
  });
}

File: modules/paralleization/tutorial_langchain.py
"""
tutorial_langchain.py

학습 목표:
1) LangChain LCEL(프롬프트 -> 모델 -> 파서) 조합 이해
2) RunnableParallel로 병렬 작업 구성
3) 병렬 결과를 구조화된 JSON으로 반환하는 흐름 학습
"""

import asyncio
import json
import os
from operator import itemgetter
from typing import Any, Final

from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import Runnable, RunnableParallel
from langchain_openai import ChatOpenAI

from tutorial_types import StepLogger, TutorialStepResult

STEP_ID: Final[str] = "step-2-langchain"
LIBRARY: Final[str] = "langchain"
OPENAI_MODEL: Final[str] = "gpt-4o-mini"
SYNTHETIC_MODEL: Final[str] = "hf:moonshotai/Kimi-K2.5"
SYNTHETIC_BASE_URL: Final[str] = "https://api.synthetic.new/openai/v1"
TUTORIAL_TOPIC: Final[str] = "Artificial Intelligence in Healthcare"


def _log_step(log: StepLogger, step_number: int, title: str) -> None:
    """공통 STEP 로그 포맷."""
    log(f"[langchain] STEP {step_number}/3 {title}")


def _resolve_model_config() -> tuple[str, str, str | None] | None:
    """키 입력은 최소화하고 모델/URL은 학습용 상수로 고정합니다."""
    openai_api_key: str = os.getenv("OPENAI_API_KEY", "").strip()
    if openai_api_key:
        return openai_api_key, OPENAI_MODEL, None

    synthetic_api_key: str = os.getenv("SYNTHETIC_API_KEY", "").strip()
    if synthetic_api_key:
        return synthetic_api_key, SYNTHETIC_MODEL, SYNTHETIC_BASE_URL

    return None


def _build_parallel_chain(llm: ChatOpenAI) -> Runnable[Any, dict[str, str]]:
    """요약/질문/키워드 체인을 병렬로 실행하는 runnable을 만듭니다."""
    summarize_chain: Runnable[Any, str] = (
        ChatPromptTemplate.from_messages(
            [
                ("system", "Summarize the following topic in one sentence."),
                ("user", "{topic}"),
            ]
        )
        | llm
        | StrOutputParser()
    )

    questions_chain: Runnable[Any, str] = (
        ChatPromptTemplate.from_messages(
            [
                ("system", "Generate two interesting questions for this topic."),
                ("user", "{topic}"),
            ]
        )
        | llm
        | StrOutputParser()
    )

    keywords_chain: Runnable[Any, str] = (
        ChatPromptTemplate.from_messages(
            [
                ("system", "Extract three key terms as comma-separated text."),
                ("user", "{topic}"),
            ]
        )
        | llm
        | StrOutputParser()
    )

    parallel_chain: RunnableParallel = RunnableParallel(
        {
            "summary": summarize_chain,
            "questions": questions_chain,
            "keywords": keywords_chain,
            "topic": itemgetter("topic"),
        }
    )
    return parallel_chain


async def run_langchain_tutorial(log: StepLogger = print) -> TutorialStepResult:
    """LangChain 단계별 튜토리얼 실행."""
    # Step 1) 실행 준비
    _log_step(log, 1, "setup")
    model_config: tuple[str, str, str | None] | None = _resolve_model_config()
    if model_config is None:
        return TutorialStepResult(
            step_id=STEP_ID,
            library="langchain",
            status="skipped",
            message="OPENAI_API_KEY or SYNTHETIC_API_KEY is missing.",
        )

    api_key: str
    model_name: str
    base_url: str | None
    api_key, model_name, base_url = model_config

    llm: ChatOpenAI = ChatOpenAI(
        model=model_name,
        temperature=0.2,
        api_key=api_key,
        base_url=base_url,
    )

    # Step 2) 병렬 runnable 구성
    _log_step(log, 2, "build parallel runnable")
    langchain_pipeline: Runnable[Any, dict[str, str]] = _build_parallel_chain(llm)

    # Step 3) 실행 및 출력 수집
    _log_step(log, 3, "execute and collect output")
    try:
        lc_result: dict[str, str] = await langchain_pipeline.ainvoke(
            {"topic": TUTORIAL_TOPIC}
        )
        output: str = json.dumps(lc_result, indent=2, ensure_ascii=True)
        return TutorialStepResult(
            step_id=STEP_ID,
            library="langchain",
            status="completed",
            message="LangChain tutorial completed.",
            output=output,
        )
    except Exception as exception:  # noqa: BLE001
        return TutorialStepResult(
            step_id=STEP_ID,
            library="langchain",
            status="failed",
            message=str(exception),
        )


async def _main() -> None:
    """직접 실행 시 단일 step 결과를 JSON으로 출력합니다."""
    result: TutorialStepResult = await run_langchain_tutorial()
    print(json.dumps(result.__dict__, indent=2, ensure_ascii=True))


if __name__ == "__main__":
    asyncio.run(_main())
